### PR TITLE
GH-40370: [C++] Define ARROW_FORCE_INLINE for non-MSVC builds

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

Compiler macros should behave consistently (as much as possible) across differently compilers.

### What changes are included in this PR?

 - Define `ARROW_FORCE_INLINE` correctly
 - Move it from `visibility.h` to `macros.h`

### Are these changes tested?

By being the base commit in changes that I'm compiling successfully and benchmarking.

* GitHub Issue: #40370

### CODE REVIEW

**Comment by felipecrv** [comment]
@kou

**Comment by kou** [comment]
* Can we change the defined header file from `util/visibility.h` (`util/macros.h`?) because this is not related to visibility?
* Can we unify this and `GDV_FORCE_INLINE`?

**Comment by pitrou** [comment]
You should also change all uses of `ARROW_FORCE_INLINE` to the new macro, no?
(I'm also not sure why you're changing the macro name, btw)

**Comment by felipecrv** [comment]
I will implement all the suggestions.

**Comment by felipecrv** [comment]
@pitrou there are no usages. I pushed a new version that doesn't rename the macro now.

File: cpp/src/arrow/util/macros.h [hunk 1]
```diff
@@ -49,10 +49,12 @@
 #define ARROW_PREDICT_TRUE(x) (__builtin_expect(!!(x), 1))
 #define ARROW_NORETURN __attribute__((noreturn))
 #define ARROW_NOINLINE __attribute__((noinline))
+#define ARROW_FORCE_INLINE __attribute__((always_inline))
 #define ARROW_PREFETCH(addr) __builtin_prefetch(addr)
 #elif defined(_MSC_VER)
 #define ARROW_NORETURN __declspec(noreturn)
 #define ARROW_NOINLINE __declspec(noinline)
+#define ARROW_FORCE_INLINE __declspec(forceinline)
```
Comment (pitrou) [inline cpp/src/arrow/util/macros.h:57]: Also define an empty fallback below?

Comment (felipecrv) [reply cpp/src/arrow/util/macros.h:57]: Done. Interestingly `ARROW_NOINLINE` was missing the empty fallback.

**Comment by felipecrv** [comment]
Fixing the Gandiva issues with the help of CI.

**Comment by felipecrv** [comment]
I'm reverting all the gandiva changes. I can't possible guess the original intentions of these annotations in there.

**Comment by felipecrv** [comment]
Dealing with Gandiva here on #40390 instead. I will rebase and create a proper issue once this one is merged

**Review by kou — APPROVED** [review]
+1

**Pull request merged**